File: frontend/src/features/tasks/services/categoryService.ts
/**
 * Category Service
 * API client for category management
 */
import type {
  Category,
  CategoryCreate,
  CategoryUpdate,
  CategoryListResponse,
} from '../types/category.types'

import { API_ENDPOINTS } from '@/shared/constants'
import { apiClient } from '@/shared/utils/api'

class CategoryService {
  async getCategories(): Promise<CategoryListResponse | null> {
    try {
      const response = await apiClient.get<CategoryListResponse>(
        API_ENDPOINTS.CATEGORIES.BASE
      )
      return response.data
    } catch (error) {
      console.error('Error fetching categories:', error)
      throw error
    }
  }

  async getCategory(id: string): Promise<Category | null> {
    try {
      const response = await apiClient.get<Category>(
        API_ENDPOINTS.CATEGORIES.BY_ID(id)
      )
      return response.data
    } catch (error) {
      console.error('Error fetching category:', error)
      throw error
    }
  }

  async createCategory(data: CategoryCreate): Promise<Category | null> {
    try {
      const response = await apiClient.post<Category>(
        API_ENDPOINTS.CATEGORIES.BASE,
        data
      )
      return response.data
    } catch (error) {
      console.error('Error creating category:', error)
      throw error
    }
  }

  async updateCategory(id: string, data: CategoryUpdate): Promise<Category | null> {
    try {
      const response = await apiClient.put<Category>(
        API_ENDPOINTS.CATEGORIES.BY_ID(id),
        data
      )
      return response.data
    } catch (error) {
      console.error('Error updating category:', error)
      throw error
    }
  }

  async deleteCategory(id: string): Promise<boolean> {
    try {
      await apiClient.delete(
        API_ENDPOINTS.CATEGORIES.BY_ID(id)
      )
      return true
    } catch (error) {
      console.error('Error deleting category:', error)
      throw error
    }
  }
}

export const categoryService = new CategoryService()

File: backend/app/services/category_service.py
"""
Category Service
Business logic for category management
"""
from sqlalchemy.orm import Session
from typing import Optional, List
from uuid import UUID

from app.repositories.category_repo import CategoryRepository
from app.schemas.category import CategoryCreate, CategoryUpdate, CategoryResponse, CategoryListResponse
from app.services.default_data_service import DefaultDataService


class CategoryService:
    """Service for Category business logic"""
    
    def __init__(self, db: Session):
        self.repo = CategoryRepository(db)
        self.db = db
    
    def create_category(self, user_id: UUID, category_data: CategoryCreate) -> CategoryResponse:
        """Create a new category"""
        category = self.repo.create(user_id, category_data)
        return CategoryResponse.model_validate(category)
    
    def get_category(self, category_id: UUID, user_id: UUID) -> Optional[CategoryResponse]:
        """Get category by ID"""
        category = self.repo.get_by_id(category_id, user_id)
        if not category:
            return None
        return CategoryResponse.model_validate(category)
    
    def get_categories(self, user_id: UUID) -> CategoryListResponse:
        """Get all categories for user (creates defaults if none exist)"""
        categories = self.repo.get_all(user_id)
        
        # Lazy init: Create default categories if user has none
        if not categories:
            default_service = DefaultDataService(self.db)
            default_service.create_default_categories(user_id)
            categories = self.repo.get_all(user_id)
        
        category_responses = [CategoryResponse.model_validate(cat) for cat in categories]
        
        return CategoryListResponse(
            categories=category_responses,
            total=len(category_responses)
        )
    
    def update_category(
        self, 
        category_id: UUID, 
        user_id: UUID, 
        category_data: CategoryUpdate
    ) -> Optional[CategoryResponse]:
        """Update category"""
        category = self.repo.update(category_id, user_id, category_data)
        if not category:
            return None
        return CategoryResponse.model_validate(category)
    
    def delete_category(self, category_id: UUID, user_id: UUID) -> bool:
        """Delete category"""
        return self.repo.delete(category_id, user_id)

